specs/core/1_beacon-chain-misc.md

```diff
@@ -0,0 +1,242 @@
+    - [Configuration](#configuration)
+    - [Containers](#containers)
+        - [`CompactCommittee`](#compactcommittee)
+        - [`ShardReceiptProof`](#shardreceiptproof)
+    - [Helper functions](#helper-functions)
+        - [`pack_compact_validator`](#pack_compact_validator)
+        - [`unpack_compact_validator`](#unpack_compact_validator)
+        - [`committee_to_compact_committee`](#committee_to_compact_committee)
+        - [`get_previous_power_of_2`](#get_previous_power_of_2)
+        - [`verify_merkle_proof`](#verify_merkle_proof)
+        - [`compute_historical_state_generalized_index`](#compute_historical_state_generalized_index)
+        - [`get_generalized_index_of_crosslink_header`](#get_generalized_index_of_crosslink_header)
+        - [`process_shard_receipt_proof`](#process_shard_receipt_proof)
+    - [Changes](#changes)
+        - [Phase 0 container updates](#phase-0-container-updates)
+            - [`BeaconState`](#beaconstate)
+            - [`BeaconBlockBody`](#beaconblockbody)
+        - [Persistent committees](#persistent-committees)
+        - [Shard receipt processing](#shard-receipt-processing)
+
+<!-- /TOC -->
+
+## Configuration
+
+| Name | Value | Unit | Duration
+| - | - | - | - |
+| `MAX_SHARD_RECEIPT_PROOFS` | `2**0` (= 1) | - | - |
+| `PERIOD_COMMITTEE_ROOT_LENGTH` | `2**8` (= 256) | periods | ~9 months |
+| `MINOR_REWARD_QUOTIENT` | `2**8` (=256) | - | - |
+
+## Containers
+
+#### `CompactCommittee`
+
+```python
+class CompactCommittee(Container):
+    pubkeys: List[BLSPubkey, MAX_VALIDATORS_PER_COMMITTEE]
+    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
+```
+
+#### `ShardReceiptProof`
+
+```python
+class ShardReceiptProof(Container):
+    shard: Shard
+    proof: List[Hash, PLACEHOLDER]
+    receipt: List[ShardReceiptDelta, PLACEHOLDER]
+```
+
+## Helper functions
+
+#### `pack_compact_validator`
+
+```python
+def pack_compact_validator(index: int, slashed: bool, balance_in_increments: int) -> int:
+    """
+    Creates a compact validator object representing index, slashed status, and compressed balance.
+    Takes as input balance-in-increments (// EFFECTIVE_BALANCE_INCREMENT) to preserve symmetry with
+    the unpacking function.
+    """
+    return (index << 16) + (slashed << 15) + balance_in_increments
+```
+
+#### `unpack_compact_validator`
+
+```python
+def unpack_compact_validator(compact_validator: int) -> Tuple[int, bool, int]:
+    """
+    Returns validator index, slashed, balance // EFFECTIVE_BALANCE_INCREMENT
+    """
+    return compact_validator >> 16, bool((compact_validator >> 15) % 2), compact_validator & (2**15 - 1)
+```
+
+#### `committee_to_compact_committee`
+
+```python
+def committee_to_compact_committee(state: BeaconState, committee: Sequence[ValidatorIndex]) -> CompactCommittee:
+    """
+    Given a state and a list of validator indices, outputs the CompactCommittee representing them.
+    """
+    validators = [state.validators[i] for i in committee]
```

Comment:
```suggestion
    assert len(committee) <= MAX_VALIDATORS_PER_COMMITTEE
    validators = [state.validators[i] for i in committee]
```

Reply:
Is this needed? This function could theoretically be used for different kinds of committees; doesn't seem wise to put a maximum designed around a specific type of committee.

Reply:
It's a safety check since the size of `CompactCommittee` elements are set to maximum `MAX_VALIDATORS_PER_COMMITTEE`:
```python
class CompactCommittee(Container):
    pubkeys: List[BLSPubkey, MAX_VALIDATORS_PER_COMMITTEE]
    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
```
It leads to one more implicit config validity condition of that `MAX_VALIDATORS_PER_COMMITTEE >= TARGET_PERSISTENT_COMMITTEE_SIZE`. (p.s. #407)

If we want to make this function could be used for different kinds of committees, how about refactoring this function into:
```python
def pack_compact_validators(state: BeaconState, committee: Sequence[ValidatorIndex]) -> Tuple[Sequence[BLSPubkey], Sequence[int]]:
    validators = [state.validators[i] for i in committee]
    compact_validators = [
        pack_compact_validator(i, v.slashed, v.effective_balance // EFFECTIVE_BALANCE_INCREMENT)
        for i, v in zip(committee, validators)
    ]
    pubkeys = [v.pubkey for v in validators]
    return pubkeys, compact_validators


def committee_to_compact_committee(state: BeaconState, committee: Sequence[ValidatorIndex]) -> CompactCommittee:
    assert len(committee) <= MAX_VALIDATORS_PER_COMMITTEE
    pubkeys, compact_validators = pack_compact_validators(state, committee)
    return CompactCommittee(pubkeys=pubkeys, compact_validators=compact_validators)
```